File: matlab/ETC.m
function etc = ETC(alg, algDataGen, baseSize, numTest, repeat)
% a data-driven model-free and parameter-free method to account for the
% factors influencing an algorithm; for instance, the crosstalk between
% algorithm realization, compilation and hardware implementation. This can
% be used to guide code optimization and to gain algorithmic maximum
% efficiency for specific hardware, programming language or compiler. The
% values of ETC in function of an input variable form a curve that offers a
% visual representation of how an algorithm scale with input size. When the
% ETC curves of different versions of an algorithm are reported together
% with the theoretical time complexity curve, their comparison allows to
% select what versions are closer to the theoretical complexity.

% alg: Provide the algorithm to test for empirical scalability.
% algDataGen: Provide the algorithm to generate data for the tested algorithm with a argument of a single size.
% baseSize: The smallest dimensional size.
% numTest: Number of different input sizes to test runtimes and ETCs.
% repeat: Number of repetition for each input size.

Sizes = baseSize*2.^((1:numTest)-1);
for rp = 1:repeat + 1
    for k = 1:numTest
        data = algDataGen(Sizes(k));
        tic
        alg(data{:});
        t(rp,k) = toc;
    end
end
t = t(2:end,:);

etc = t ./ t(:,1);


%% plot
MAG=1;
f1=figure(1);
set(f1,'Resize',true)
f1.Units='centimeters';
f1.Position(3:4)=[18,8]*MAG;

options.handle     = figure(1);
options.alpha      = 0.5;
options.line_width = 0.75;
options.error      = 'std'; 
options.x_axis     = Sizes/Sizes(1);
alpha=0.45;
sa1 = subplot(1,2,1);

options.color_line = [0 0 1]*0.9;
options.color_area = options.color_line * (1-alpha) +[1 1 1]*alpha;
error_curve(etc,options);
plot(options.x_axis, mean(etc,1),'o','LineWidth',1.25,'MarkerSize',5,'Color',options.color_line,'Tag','point')

sa1.XScale='log';
sa1.YScale='log';
xlabel({'Number of times as the smallest input size'})
ylabel("{ETC}")

plot([1,Sizes(end)/Sizes(1)],[1,Sizes(end)/Sizes(1)],'k')
plot([1,Sizes(end)/Sizes(1)],[1,Sizes(end)/Sizes(1)].^2,'k')
plot([1,Sizes(end)/Sizes(1)],[1,Sizes(end)/Sizes(1)].^3,'k')


options.handle     = figure(1);
options.alpha      = 0.5;
options.line_width = 0.75;
options.error      = 'std'; 
options.x_axis     = Sizes;
alpha=0.45;
sa2 = subplot(1,2,2);

options.color_line = [0 0 1]*0.9;
options.color_area = options.color_line * (1-alpha) +[1 1 1]*alpha;
error_curve(t,options);
plot(Sizes,t,'o','LineWidth',1.25,'MarkerSize',5,'Color',options.color_line,'Tag','point')

sa2.XScale='log';
sa2.YScale='log';
xlabel('Input data dimensional size')
ylabel("{ Running time} {\it t} (s)")

end

File: python/ETC.py
import numpy as np
import matplotlib.pyplot as plt
import time

def ETC(alg, alg_data_gen, base_size, num_test, repeat):
    """
    A data-driven, model-free, and parameter-free method to evaluate how an algorithm scales
    with input size on specific hardware and configurations.

    Parameters:
    alg: Callable - Algorithm to test for empirical scalability.
    alg_data_gen: Callable - Function to generate input data for `alg` with a single size argument.
    base_size: int - The smallest dimensional size.
    num_test: int - Number of different input sizes to test.
    repeat: int - Number of repetitions for each input size.

    Returns:
    etc: np.ndarray - The Empirical Time Complexity (ETC) matrix.
    """
    # Generate input sizes
    sizes = base_size * 2 ** np.arange(num_test)
    runtimes = np.zeros((repeat+1, num_test))

    # Measure runtime for each size and repetition
    for rp in range(repeat+1):
        for k, size in enumerate(sizes):
            data = alg_data_gen(size)
            start_time = time.perf_counter()
            alg(*data)
            runtimes[rp, k] = time.perf_counter() - start_time
    runtimes = runtimes[1:][:]
    # Compute ETC as normalized runtimes
    etc = runtimes / runtimes[:, 0][:, np.newaxis]

    # Plot results
    plot_etc(sizes, etc, runtimes)

    return etc

def plot_etc(sizes, etc, runtimes):
    """
    Plot the Empirical Time Complexity (ETC) and runtime results.

    Parameters:
    sizes: np.ndarray - Array of input sizes.
    etc: np.ndarray - ETC matrix.
    runtimes: np.ndarray - Raw runtimes matrix.
    """
    x_axis_etc = sizes / sizes[0]
    x_axis_runtime = sizes

    # ETC Plot
    plt.figure(figsize=(12, 6))

    plt.subplot(1, 2, 1)
    mean_etc = np.mean(etc, axis=0)
    std_etc = np.std(etc, axis=0)
    plt.fill_between(x_axis_etc, mean_etc - std_etc, mean_etc + std_etc, color='blue', alpha=0.3)
    plt.plot(x_axis_etc, mean_etc, 'o-', label='ETC', color='blue')
    plt.xscale('log')
    plt.yscale('log')
    plt.xlabel('Number of times as the smallest input size')
    plt.ylabel('ETC')
    plt.title('Empirical Time Complexity (ETC)')
    plt.grid(True, which="both", linestyle="--", linewidth=0.5)
    plt.legend()

    # Reference lines for theoretical complexities
    plt.plot(x_axis_etc, x_axis_etc, 'k--', label='O(n)')
    plt.plot(x_axis_etc, x_axis_etc**2, 'k--', label='O(n^2)')
    plt.plot(x_axis_etc, x_axis_etc**3, 'k--.', label='O(n^3)')


    # Runtime Plot
    plt.subplot(1, 2, 2)
    mean_runtimes = np.mean(runtimes, axis=0)
    std_runtimes = np.std(runtimes, axis=0)
    plt.fill_between(x_axis_runtime, mean_runtimes - std_runtimes, mean_runtimes + std_runtimes, color='blue', alpha=0.3)
    plt.plot(x_axis_runtime, mean_runtimes, 'o-', label='Runtime', color='blue')
    plt.xscale('log')
    plt.yscale('log')
    plt.xlabel('Input data dimensional size')
    plt.ylabel('Running time (s)')
    plt.title('Running Time')
    plt.grid(True, which="both", linestyle="--", linewidth=0.5)
    plt.legend()

    plt.tight_layout()
    plt.show()
